src/main/kotlin/org/rust/ide/inspections/checkMatch/CheckMatchUtils.kt

```diff
@@ -0,0 +1,159 @@
+ */
+
+package org.rust.ide.inspections.checkMatch
+
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.types.ty.Ty
+import org.rust.lang.core.types.ty.TyAdt
+import org.rust.lang.core.types.ty.TyTuple
+import org.rust.lang.core.types.ty.TyUnknown
+import org.rust.lang.core.types.type
+
+private val RsEnumVariant.index: Int
+    get() = parentEnum.variants.indexOf(this)
+
+typealias Matrix = List<List<Pattern>>
+
+val Matrix.type: Ty
+    get() = flatten().map { it.ty }.find { it !is TyUnknown } ?: TyUnknown
+
+val RsMatchExpr.matrix: Matrix
+    get() = arms.flatMap { arm -> arm.patList.map { listOf(it.lower) } }
+
+typealias FieldPattern = Pair<Int, Pattern>
+
+private val RsExpr.value: Constant
+    get() = when (this) {
+        is RsLitExpr -> {
+            val kind = kind
+            when (kind) {
+                is RsLiteralKind.Boolean -> Constant.Boolean(kind.value)
+                is RsLiteralKind.Integer -> Constant.Integer(kind.value ?: 0)
+                is RsLiteralKind.Float -> Constant.Double(kind.value ?: 0.0)
+                is RsLiteralKind.String -> Constant.String(kind.value ?: "")
+                is RsLiteralKind.Char -> Constant.Char(kind.value ?: "")
+                null -> Constant.Unknown
+            }
+        }
+        is RsPathExpr -> Constant.Path(this)
+        is RsUnaryExpr -> TODO()
+        else -> TODO()
+    }
+
+private val RsPat.type: Ty
+    get() = when (this) {
+        is RsPatConst -> expr.type
+        is RsPatStruct, is RsPatTupleStruct -> {
+            val path = (this as? RsPatTupleStruct)?.path ?: (this as RsPatStruct).path
+            val resolved = path.reference.resolve()
+            when (resolved) {
+                is RsEnumVariant -> TyAdt.valueOf(resolved.parentEnum)
+                is RsStructOrEnumItemElement -> TyAdt.valueOf(resolved)
+                else -> TyUnknown
+            }
+        }
+        is RsPatWild -> TyUnknown
+        is RsPatIdent -> patBinding.type
+        is RsPatTup -> TyTuple(patList.map { it.type })
+        is RsPatRange -> patConstList.firstOrNull()?.type ?: TyUnknown
+
+        is RsPatRef -> TODO()
+        is RsPatMacro -> TODO()
+        is RsPatSlice -> TODO()
+        else -> TODO()
+    }
+
+// lower_pattern_unadjusted
+private val RsPat.kind: PatternKind
+    get() = when (this) {
+        is RsPatIdent -> {
+            val resolved = patBinding.reference.resolve()
+            if (resolved is RsEnumVariant) {
+                PatternKind.Variant(resolved.parentEnum, resolved.index, emptyList())
+            } else {
+                PatternKind.Binding(patBinding.type, patBinding.name.orEmpty())
+            }
+        }
+
+        is RsPatWild -> PatternKind.Wild
+
+        is RsPatTup -> PatternKind.Leaf(patList.mapIndexed { i, pat -> i to pat.lower })
+
+        is RsPatStruct -> {
+            val item = path.reference.resolve() ?: error("Can't resolve ${path.text}")
```

Comment:
Can we rewrite this in a more error-tolerant way? In this PR almost all errors are thrown in non-critical situations, imo.

Reply:
I strongly agree with you. What do you think about throwing exceptions and catch it inside `RsMatchCheckInspection.visitMatchExpr`? IMO exceptions are suitable there because we want to stop the match checking and don't provide any errors/fixes if there are some unresolved stuff inside.

Reply:
Java exceptions are too heavyweight to use for the normal flow of control, especially in inspections, IMO. @vlad20012 @Undin what do you think?

Reply:
I think it's mostly OK except the fact the we don't have checked exceptions in Kotlin and can miss to handle it =/